File: tests/test_invariants.py
import unittest
import os
import sys

# Add src to path so we can import
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from src.verifier import verifyAction, PreconditionError, InvariantViolationError

class TestInvariants(unittest.TestCase):
    def setUp(self):
        self.dsl = {
            'state_schema': {'balance': 'Number'},
            'invariants': [
                {'type': 'NEVER', 'condition': 'balance < 0'}
            ],
            'actions': {
                'Withdraw': {
                    'requires': ['balance >= 50'],
                    'mutates': ['balance = balance - 100']
                }
            }
        }

    def test_precondition_error(self):
        # Initial balance is 30, which fails the 'balance >= 50' requirement
        current_state = {'balance': 30}
        
        with self.assertRaises(PreconditionError) as context:
            verifyAction(self.dsl, current_state, 'Withdraw')
            
        self.assertIn('Precondition failed', str(context.exception))

    def test_invariant_violation_error(self):
        # Initial balance is 50, passes precondition.
        # Mutation subtracts 100, balance becomes -50.
        # Violates 'NEVER: balance < 0'
        current_state = {'balance': 50}
        
        with self.assertRaises(InvariantViolationError) as context:
            verifyAction(self.dsl, current_state, 'Withdraw')
            
        self.assertIn('InvariantViolationError', str(context.exception))
        self.assertIn('NEVER balance < 0', str(context.exception))

if __name__ == '__main__':
    unittest.main()

File: tests/test_invariants.ts
import { verifyAction, PreconditionError, InvariantViolationError } from '../src/verifier';

function runTests() {
  console.log('Running test_invariants.ts...');
  
  const dsl = {
    state_schema: { balance: 'Number' },
    invariants: [
      { type: 'NEVER', condition: 'balance < 0' }
    ],
    actions: {
      Withdraw: {
        requires: ['balance >= 50'],
        mutates: ['balance = balance - 100']
      }
    }
  };

  // Test 1: PreconditionError
  try {
    verifyAction(dsl, { balance: 30 }, 'Withdraw');
    throw new Error('Expected PreconditionError but no error was thrown');
  } catch (e: any) {
    if (e instanceof PreconditionError) {
      console.log('✅ PreconditionError correctly raised for failed precondition');
    } else {
      throw new Error(`Expected PreconditionError, got ${e.name}`);
    }
  }

  // Test 2: InvariantViolationError
  try {
    // Precondition passes (balance >= 50), but mutation (balance - 100) results in -50
    // This violates the NEVER: balance < 0 invariant
    verifyAction(dsl, { balance: 50 }, 'Withdraw');
    throw new Error('Expected InvariantViolationError but no error was thrown');
  } catch (e: any) {
    if (e instanceof InvariantViolationError) {
      console.log('✅ InvariantViolationError correctly raised for NEVER invariant');
    } else {
      throw new Error(`Expected InvariantViolationError, got ${e.name}`);
    }
  }
  
  console.log('✅ test_invariants.ts passed!');
}

runTests();
